Bird Sounds Global (Kerttu Global) project

aorin opened the pull request:
Can be tested: https://kerttu-global-dev.laji.fi/
Laji.fi can be tested (shouldn't have any visible changes): https://177852929.dev.laji.fi/

This pull request adds a folder `projects/kerttu-global` that contains the source code for Bird Sounds Global site.

There are some changes to shared files:
- Kerttu audio viewer is refactored from `projects/laji/src/app/+theme/kerttu/audio-viewer` to `projects/laji/src/app/shared-modules/audio-viewer` and new features are added
- changes to laji-datatable `projects/laji/src/app/shared-modules/datatable/*`
    - an option `striped` that determines whether the table is striped or not is added to the datatable
    - a new datatable cell template is added, `booleanCheck`
    - two new classes are added: `.red-row`, `.yellow-row`
- an ability to set different default language than Finnish, changes are in these files:
    - `projects/laji/src/app/locale/localize.guard.ts`
    - `projects/laji/src/app/locale/localize-router.service.ts`

olzraiti commented on projects/laji/src/app/shared-modules/audio-viewer/audio-viewer/audio-spectrogram/audio-spectrogram.component.ts:
```diff
@@ -0,0 +1,105 @@
+  }
+
+  ngAfterViewInit() {
+    // update width and height immediately and after some time has passed to ensure that the content has loaded
+    timer(0, 300).pipe(take(2)).subscribe(() => {
+      this.onResize();
+      this.cdr.markForCheck();
+    });
```
Is the timer really necessary? I tried removing it and the spectogram view initialized just fine. Is there some race condition that occurs on slower machines?

aorin replied:
It's necessary, `this.containerRef.nativeElement.offsetWidth` is not always correct at `ngAfterViewInit`. It happens more frequently when you change between views (double click some template/candidate and then press cancel).

olzraiti replied:
How about updating on `laji-spectogram` on `spectogramReady`? I tested at seems to work. Btw, the `ngOnChanges` stuff should maybe be done on `ngAfterViewChecked`.

aorin replied:
It may work because `laji-spectrogram` has a delay before it emits `spectrogramReady` (line 60: `observable.pipe(delay(0))`). It has that because sometimes the spectrogram is ready straight away and you couldn't emit an output on the `ngOnChanges` function (not sure if it's fixed in the current Angular version). I wouldn't add it to that because it may be confusing that it depends on another component.

> Btw, the ngOnChanges stuff should maybe be done on ngAfterViewChecked.

I think `ngOnChanges` is ok place because it reacts to changes in `Input` variables. I did small refactoring for the component [here](https://github.com/luomus/laji/pull/219/commits/a6baa4331b8cce67672169e4f6f7ec0f83d1185a) to make it hopefully little clearer

olzraiti replied:
Hmm. I didn't notice the `delay(0)`. I tried to implement this without the delay and the `timer`, and rely on hooks (`AfterViewChecked`) but didn't get it to work. The canvas & img drawing should be synchronous browser stuff, so not sure what's going on there. Well, I've spent enough time on this now - if it works it works... :D if you know why the `delay(0)` is needed, could you please make a comment for that?

aorin replied:
I added a comment but I'm not sure what is the reason for that. I remember it worked ok without SSR before and only SSR was causing problems. Now when I tested it doesn't work even without SSR if there is no delay...

olzraiti commented on projects/laji/src/app/shared-modules/audio-viewer/audio-viewer/audio-viewer.component.ts:
```diff
@@ -0,0 +1,238 @@
+    } else {
+      this.clicks++;
+      if (this.clicks === 1) {
+        setTimeout(() => {
```
Why the timeout?

aorin replied:
It's a method for detecting that the click is a single click (and not a double-click). It waits a while and if the number of clicks have not increased, the click is determined to be a single click. I added some comments about this to the code.

olzraiti commented on projects/kerttu-global/src/app/+validation/service/query-reset.guard.ts:
```diff
@@ -0,0 +1,15 @@
+import { SpeciesListQueryService } from './species-list-query.service';
+
+@Injectable()
+export class QueryResetGuard implements CanActivate {
```
The name is a bit generic, how about `SpeciesListQueryResetGuard`? That would help understanding the connection with the species list query  service.

olzraiti reviewed (changes requested):
Overall it was pretty straightforward to understand, good work! I added some minor nits, and I'd like to hear more about the reasoning behind some timers used.

The modules could be tidied up a bit: audio viewer module could be a non-shared module and the kerttu stuff under theme also, so they could be lazy loaded, which makes laji.fi less bloated. What do you think? Doesn't have to be implemented in this PR necessarily.

olzraiti commented:
Stumbled on a bug:

1. Select 1st order (struthioniformes)
2. Select 1st family (struthionidae)
3. Select some other order
4. Clear order (select the empty option)
-> struthionidae appears as family

I guess the family should stay cleared?

olzraiti commented:
One thing that I forgot to comment: could we use a different component prefix name for kerttu project than `laji` to prevent namespace pollution?

aorin commented on projects/kerttu-global/src/app/+validation/species-list/species-list.component.ts:
```diff
@@ -0,0 +1,55 @@
+
+  private changeQuery() {
+    this.queryChange.emit(this.query);
+  }
```
Does [this](https://github.com/luomus/laji/pull/219/commits/b441a41a2efa7bdcb187eadcb734bef8b98c0de5) look better?

olzraiti replied:
Yes! :+1:

aorin commented on projects/kerttu-global/src/app/+validation/species-list/species-table/species-table.component.ts:
```diff
@@ -0,0 +1,81 @@
+  @Input() sorts: DatatableSort[] = [];
+
+  private defaultColumns: DatatableColumn[] = [];
+  columns: DatatableColumn[] = [];
```
Removed unnecessary `defaultColumns`. Can't initialize it before `ngOnInit` because `this.notificationTpl` is undefined before that

olzraiti replied:
:+1:

aorin commented on projects/kerttu-global/src/app/+validation/validation.component.ts:
```diff
@@ -0,0 +1,76 @@
+    ).subscribe(data => {
+      this.speciesList = data;
+      this.loading = false;
+      this.cd.markForCheck();
```
Added `changeDetection: ChangeDetectionStrategy.OnPush` that was missing for the component (not sure if that was what you meant)

olzraiti replied:
Partially, yes - when you use `changeDetection: ChangeDetectionStrategy.OnPush` you don't have to manually use `this.cd.markForCheck();` to update the view if you update the instance variables immutably (like is done here)

olzraiti replied:
Of course calling `this.cd.markForCheck()` doesn't do much harm, and could be argued to be OK as "safe to be safe than sorry", but at least for me it's a code smell:

* though it doesn't do much harm performance wise, it still does some. None < less.
* When I read components, if it uses the OnPush strategy and there are no direct change detection checking, I feel safe that the component is coded cleanly as it works in an immutable way and I can edit the component with more confidence knowing that if I continue working on the component in a clean, immutable way, it won't break.

aorin replied:
The `this.cd.markForCheck()` is still needed here because it's inside RxJS `subscribe` and the view wouldn't update immediately without it

olzraiti replied:
Argh! Thanks, you're right. Hmm, maybe using the `async` pipe would be an improvement, but it's fine as it is now :)

aorin commented on projects/kerttu-global/src/app/kerttu-global-routing.module.ts:
```diff
@@ -0,0 +1,67 @@
+@Injectable()
+export class PreloadSelectedModulesList implements PreloadingStrategy {
+  preload(route: Route, load: () => Observable<any>): Observable<any> {
+    const loadRoute = () => ObservableTimer(50).pipe(flatMap(() => load()), catchError(() => ObservableOf(null)));
```
I did this the same way than in the `IUCN` project but it seems that it's a quite common practice to add some delay when preloading modules, to prevent network bottleneck

olzraiti replied:
Oh, I did some reading on this, neat!

aorin commented on projects/kerttu-global/src/environments/global.ts:
```diff
@@ -0,0 +1,53 @@
+  sources: {
+    kotka: 'KE.3'
+  }
+};
```
Removed this file and imported `type.kerttuGlobal` from laji globals

aorin commented:
Thanks for all the comments! I fixed the bug and changed the prefix of the project to `bsg`.

> The modules could be tidied up a bit: audio viewer module could be a non-shared module and the kerttu stuff under theme also, so they could be lazy loaded, which makes laji.fi less bloated. What do you think? Doesn't have to be implemented in this PR necessarily.

I agree that the Kerttu could be refactored to its own module and lazy loaded but is there a problem with audio viewer being in shared modules? Isn't it loaded only when it is imported?

olzraiti commented:
> is there a problem with audio viewer being in shared modules? Isn't it loaded only when it is imported?

Sorry, I mixed up the `SharedModule` (everything under `shared` is always loaded if shared module is loaded, that is, almost always) and stuff `shared-modules` folder... So no problem, that's a good place for it!

olzraiti reviewed (approved):
:+1:
